compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -13461,454 +13461,133 @@ OMR::Z::TreeEvaluator::arraysetEvaluator(TR::Node * node, TR::CodeGenerator * cg
-                     oldDep->getPostConditions()->getRegisterDependency(i)->getRegister(cg)->decTotalUseCount();
-               }
-            TR::Instruction * s390cursor = cursor;
-            s390cursor->resetDependencyConditions();
-            s390cursor->setDependencyConditionsNoBookKeeping(loopDep);
-            loopDep->bookKeepingRegisterUses(s390cursor, cg);
-            oldS390cursor->resetDependencyConditions();
-            loopDep->setIsUsed();
-            }
+void
+OMR::Z::TreeEvaluator::referenceArraycopyEvaluator(TR::Node* node, TR::CodeGenerator* cg, TR::Node* byteSrcNode, TR::Node* byteDstNode, TR::Node* byteLenNode, TR::Node* byteSrcObjNode, TR::Node* byteDstObjNode)
+   {
+   TR::Register* byteSrcObjReg = cg->evaluate(byteSrcObjNode);
+   TR::Register* byteDstObjReg = cg->evaluate(byteDstObjNode);
 
-         vtOp.cleanUpReg();
-         }
-      }
+   // For reference array copies that need an array store check always call the helper function
+   if (!node->isNoArrayStoreCheckArrayCopy())
+      {
+      TR::Register* byteSrcReg = cg->evaluate(byteSrcNode);
+      TR::Register* byteDstReg = cg->evaluate(byteDstNode);
+      TR::Register* byteLenReg = cg->evaluate(byteLenNode);
 
 #ifdef J9_PROJECT_SPECIFIC
-   if (!isSimpleCopy)
-      {
-      TR::TreeEvaluator::genWrtbarForArrayCopy(node, cg->evaluate(byteSrcObjNode),
-                            cg->evaluate(byteDstObjNode), byteSrcNode->isNonNull(), cg);
-      }
+      TR::TreeEvaluator::genArrayCopyWithArrayStoreCHK(node, byteSrcObjReg, byteDstObjReg, byteSrcReg, byteDstReg, byteLenReg, cg);
 #endif
 
-   if (byteLenReg) cg->stopUsingRegister(byteLenReg);
-
-   if (isSimpleCopy)
-      {
-      cg->decReferenceCount(byteLenNode);
-
-      if (evaluateChildren)
-         {
-         cg->decReferenceCount(byteDstNode);
-         cg->decReferenceCount(byteSrcNode);
-         }
-      }
-   else
-      {
-      cg->decReferenceCount(byteSrcObjNode);
-      cg->decReferenceCount(byteDstObjNode);
       cg->decReferenceCount(byteSrcNode);
       cg->decReferenceCount(byteDstNode);
       cg->decReferenceCount(byteLenNode);
       }
-
-   //
-   // If we made copies of the derived registers, these copies are now dead.
-   // At this point, copyByte{Src,Dst,Len}Reg == byte{Src,Dst,Len}Reg, respectively.
-   //
-   if (copyByteSrcReg)
-      {
-      cg->stopUsingRegister(byteSrcReg);
-      }
-   if (copyByteDstReg)
-      {
-      cg->stopUsingRegister(byteDstReg);
-      }
-   if (copyByteLenReg)
+   else
       {
-      cg->stopUsingRegister(byteLenReg);
+      // TODO: There are cases under which the following assert will trigger. These need to be fixed and this assert needs to be re-enabled.
+      // TR_ASSERT(node->getArrayCopyElementType() == TR::Address, "Reference arraycopy element type should be a TR::Address but was '%s'", node->getArrayCopyElementType().toString());
```

Comment:
It seems like this assert could be re-enabled now that we can only get here if it's a ref array copy but I'm not 100% sure about that. Worth making a note of this anyway.